crates/net/network/src/transactions/fetcher.rs

```diff
@@ -17,61 +21,143 @@ use tokio::sync::{mpsc::error::TrySendError, oneshot, oneshot::error::RecvError}
+/// [`DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_MESSAGE`], which defaults to 256
+/// KiB, so 128 KiB.
+const DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_FETCH_PENDING: usize =
+    1 / 2 * DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_MESSAGE;
+
+/// Average byte size of an encoded transaction. Default is
+/// [`DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_MESSAGE`], which defaults to 128
+/// KiB, divided by [`SOFT_LIMIT_COUNT_HASHES_GET_POOLED_TRANSACTIONS_REQUEST`], which defaults to
+/// 256 hashes, so 521 bytes.
+const AVERAGE_BYTE_SIZE_TX_ENCODED: usize =
+    DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_MESSAGE /
+        SOFT_LIMIT_COUNT_HASHES_GET_POOLED_TRANSACTIONS_REQUEST;
+
+/// Median observed size in bytes of a small encoded legacy transaction. Default is 120 bytes.
+const MEDIAN_BYTE_SIZE_SMALL_LEGACY_TX_ENCODED: usize = 120;
+
+/// Default budget for finding an idle fallback peer for any hash pending fetch, when said search
+/// is budget constrained. Default is a sixth of [`DEFAULT_MAX_HASHES_PENDING_FETCH`], which
+/// defaults to 12 800 hashes (the breadth of the search), divided by
+/// [`DEFAULT_MAX_COUNT_FALLBACK_PEERS_PER_TX`], which defaults to 3 peers (the depth of the
+/// search), so 4 266 hashes.
+const DEFAULT_BUDGET_FIND_IDLE_FALLBACK_PEER: usize =
+    1 / 6 * DEFAULT_MAX_HASHES_PENDING_FETCH / DEFAULT_MAX_COUNT_FALLBACK_PEERS_PER_TX as usize;
 
 /// The type responsible for fetching missing transactions from peers.
 ///
 /// This will keep track of unique transaction hashes that are currently being fetched and submits
 /// new requests on announced hashes.
 #[derive(Debug)]
 #[pin_project]
-pub(super) struct TransactionFetcher {
-    /// All peers to which a request for pooled transactions is currently active. Maps 1-1 to
-    /// `inflight_requests`.
+pub(super) struct TransactionFetcher<T = TxFetchMetadata> {
+    /// All peers with an `inflight_requests`.
     pub(super) active_peers: LruMap<PeerId, u8, ByLength>,
     /// All currently active requests for pooled transactions.
     #[pin]
     pub(super) inflight_requests: FuturesUnordered<GetPooledTxRequestFut>,
     /// Hashes that are awaiting an idle peer so they can be fetched.
-    // todo: store buffered eth68 and eth66 hashes separately
-    pub(super) buffered_hashes: LruCache<TxHash>,
-    /// Tracks all hashes that are currently being fetched or are buffered, mapping them to
-    /// request retries and last recently seen fallback peers (max one request try for any peer).
-    pub(super) unknown_hashes: LruMap<TxHash, (u8, LruCache<PeerId>), Unlimited>,
-    /// Size metadata for unknown eth68 hashes.
-    pub(super) eth68_meta: LruMap<TxHash, usize, Unlimited>,
+    pub(super) hashes_pending_fetch: LruCache<TxHash>,
+    /// Tracks all hashes in the transaction fetcher.
+    pub(super) hashes_unknown_to_pool: LruMap<TxHash, T, Unlimited>,
```

Comment:
it's not obvious what the difference here is.

> Hashes that are awaiting an idle peer so they can be fetched.

> Tracks all hashes in the transaction fetcher.

aren't those the same thing?

Reply:
from compiler development there is a very nice type in the [cranelift compiler called `PrimaryMap`](https://github.com/bytecodealliance/wasmtime/blob/2c0e528b9fad9b93e55b03abb816360f3c1bb52d/cranelift/entity/src/primary.rs#L16-L39). `PrimaryMap` is a dense index space in the sense that [all keys are sequential and are created upon pushing a value into the map](https://github.com/bytecodealliance/wasmtime/blob/de051941c98e30eb86637206fd50e3599eba0853/cranelift/entity/src/primary.rs#L121-L126). then there is a another type called [`SecondaryMap`](https://github.com/bytecodealliance/wasmtime/blob/de051941c98e30eb86637206fd50e3599eba0853/cranelift/entity/src/map.rs#L18-L35), which doesn't allocate entities, but rather uses the entities from the `PrimaryMap` as keys. It's also enforced that there can be max one `PrimaryMap` instance for a type that implements `EntityRef`. this is the pattern I wanted to follow here, `hashes_unknown_to_pool` is the "primary map" and `hashes_pending_fetch` is the "secondary map".

since we can't fit a 256-byte `TxHash` in a `usize`, we can't implement `EntityRef` for it. I wanted to use some similar types here, but didn't find a simple way. this is the case since other integrated components, like for example a session that is forwarding transactions to the tx fetcher, uses the `TxHash` to index transactions, and they all need to use the same index to avoid expensive look ups for translating one components index to another components index when interfacing between components. so using a smaller type that implements EntityRef like `u32`  to index hashes, would be a crate-wide change, maybe even workspace-wide change. it would be nice but is out of scope of this pr. then we would have a map somewhere with a `u32->TxHash` mapping, and only pass the 32-bit reference around instead of the 256-bit reference we do now. but! we can't do that since the hashes in circulation in the node always change. we can do `u32->TxHashContainer` in a fixed size store for hashes.

the store `hashes_unknown_to_pool` should be flexible sized for now, so that's off the table here. will think of better names for the fields though.